### ISSUE TITLE
FEATURE REQUEST: Log function calls from game

### ISSUE THREAD

**Issue by Templarfreak** [narzoul/DDrawCompat#11]
If it would be possible to create a log of what the game calls in order to draw it's graphics I think it would not only be much easier to debug specific problems with specific games but could also be helpful in developing the wrapper in general. 

For instance, I tried this with Red Alert 2: Yuri's Revenge and in it the menus are completely gone. If you can navigate the invisible menus you can get to the game fine and the game looks fine and is actually even faster than normal (The game runs exceptionally smooth already minus some problems that a hack called ARES already fixes), however the in-game menu is still not rendering. 

I was mainly trying this as I'm looking around for potential solutions to a bug to the game on modern hardware and systems (that somehow not everyone seems to have) where popup tooltips seem to only render for a single frame and very sporadically and at seemingly random. (Sometimes they don't render for several seconds and sometimes they render several times in a row in a single second)

The question is, would this actually be possible? This would probably be a tall order as I'd imagine there'd be no easy way to do this.

Perhaps also logging system specs and OS. Noticed that's not actually being logged in the ddraw.log.

**Comment by narzoul** [comment]
This is practically already implemented, you just have to compile in debug mode. Note that this has a heavy performance hit. If you want to have performance closer to release mode with debug logs, you can globally search and replace "#ifdef _DEBUG" with "#ifndef _DEBUG" and compile in release mode.

But there are more specialized tools available for this purpose, I recommend [API Monitor](http://www.rohitab.com/apimonitor).

**Comment by Templarfreak** [comment]
This should be something that you suggest to people to use if they experience/encounter problems and should be possible by default (without having to download a separate version) that can be enabled by putting it in some config. Like in a DDrawCompat.cfg or .ini or something and putting "Debug=yes" in it

**Comment by narzoul** [comment]
I surely wouldn't suggest the average end user to take debug logs and try to find solutions to their problems that way. The logs are only useful to developers who are familiar with the DirectDraw and Win32 APIs and even then you might have to check the source code to interpret some of the logs because the format isn't documented anywhere and not all parameters are logged completely. People capable of doing that surely would have no problem compiling a debug version either.

Perhaps you meant only so that they can send the logs to me, but in my experience most of the issues occur at later points in games, by which time the logs can have grown to hundreds of megabytes and it's not feasible to upload them.

Enabling the logs via a config setting may be possible, but it adds ~1mb to the binary size and may have a negative performance impact even when logging is disabled due to having an extra check on each function entry/exit, which currently isn't compiled in at all. If I can measure it to have no significant performance impact then I might consider adding it, but even then I might prefer a separate binary.

But before I add such an option, I would first like to understand why do you insist on doing this with DDrawCompat when there are far better tools for the job that are already implemented, have capture/display filters that cover a lot more APIs than DDrawCompat (which can only log interfaces that it already hooks for compatibility), can list all parameters and results properly formatted, can take stack traces for each call, etc. The one I mentioned above even has a nice GUI and is not at all difficult to use, have you given it a try?

**Comment by Templarfreak** [comment]
It would just be better and more convenient to people using DDrawCompat to not have to download a separate program or version of DDrawCompat in order to send you debug logs. I understand the other problems, though.

As for the size of the logs, maybe a timestamp can help there to sort through it all better? Or perhaps simply telling people to close the game as soon as the problem occurs (to narrow it down to definitely being near the bottom of the log)?

**Comment by narzoul** [comment]
Well, I did some improvements to logging so debug logs are more readable now (e.g. nested calls are indented). I've also added a new build type called "ReleaseWithDebugLogs", which as the name suggests is mostly identical to the release build, except all of the debug logs are included. I've uploaded it as a separate zip file to the experimental release. This is the best I could do with the time I had so far.
